Switched the language server from websockets to TCP

ofrank123 opened the pull request:
The new language server for GDScript is a great inclusion to the engine, but it was implemented in websockets, which is fairly unusual for language servers.  Most language servers use either a stdio connection or a TCP server to communicate, and as such these protocols are widely supported by LSP clients.  Although VSCode and Atom are currently working with the websocket implementation, other text editors like emacs (my editor of choice) and vim do not have mature enough websocket support for their respect lsp clients to work with the protocol, as was noted in [this issue](https://github.com/emacs-lsp/lsp-mode/issues/1361) on emacs' lsp-mode .  Of course this change would be breaking for the vscode and atom clients, but the code in them that handles the websocket server is minimal, and could be changed over easily.  This is a work in progress, as it cannot currently handle multiple client connections (although i don't see this as a huge issue, but I thought I'd make the pull request so that others knew it was being worked on.  It currently works with a single connection at a time, and I have tested it with emacs' lsp-mode on [my fork](https://github.com/ofrank123/lsp-mode).  @Geequlim has mentioned before that he is not opposed to switching to TCP either.

ofrank123 committed:
Migrating language server from Websockets to raw TCP

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -32,32 +32,65 @@
+				CharString charstr = output.utf8();
+				peer->put_data((const uint8_t *)charstr.ptr(), charstr.length());
+			}
+		}
```
```suggestion
		}
		memfree(content);
```

CyanBlob commented:
Thanks for fixing the leak so quickly! I'll test this with vim when I get home and review it further. Are you still going to add support for multiple clients? Is that functionality still wanted?

ofrank123 commented:
I'd be happy to go about adding it, but I can't think of a usecase where one might want to have multiple text editors connected at the same time.

CyanBlob commented:
I personally agree with you. @Geequlim, what was your initial usecase for supporting multiple clients? Do you think that functionality should be kept?

ofrank123 commented:
I should also note I've removed the notify_all_clients method, as I couldn't find it being used anywhere

ofrank123 commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -32,32 +32,66 @@
+
+		// Read bytes until end of header
+		while (!header.ends_with("\r\n\r\n")) {
+			uint8_t *buff = (uint8_t *)memalloc(1);
```
Yes I was going to change this actually

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -32,32 +32,66 @@
 
-void GDScriptLanguageProtocol::on_client_connected(int p_id, const String &p_protocal) {
-	clients.set(p_id, server->get_peer(p_id));
+void GDScriptLanguageProtocol::on_client_connected() {
```
We should store a collection of peers instead of using just one. Then, when one connects, add it to the collection. When it disconnects, remove it from the collection

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +185,45 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+	if (server->is_connection_available()) {
+		on_client_connected();
+	}
+	if (peer != NULL) {
```
If we have any peers in the collection:

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +185,45 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+		on_client_connected();
+	}
+	if (peer != NULL) {
+		StreamPeerTCP::Status status = peer->get_status();
```
Iterate over all peers

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +185,45 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+	if (peer != NULL) {
+		StreamPeerTCP::Status status = peer->get_status();
+		if (status == StreamPeerTCP::STATUS_NONE || status == StreamPeerTCP::STATUS_ERROR) {
+			on_client_disconnected();
```
Pass disconnected peer to on_client_disconnected for removal from peers container

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +185,45 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+		} else {
+			int bytes = peer->get_available_bytes();
+			if (bytes > 0) {
+				on_data_received(bytes);
```
Pass peer along with bytes, or:
Pass just the peer, reading the bytes in on_data_received (this is what I would suggest)

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +185,45 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
-	CharString charstr = msg.utf8();
 
-	(*peer)->put_packet((const uint8_t *)charstr.ptr(), charstr.length());
+	peer->put_string(msg);
```
Need to notify the correct peer(s)

CyanBlob reviewed (commented):
It looks like there's a consensus that we should continue supporting multiple clients. The TCP_Server and StreamPeerTCP objects that we're already using will make this easy.
In general, what we need to change is:
Use a container of `peer`s instead of a single peer
When polling:

-  Add new peers to the container
- Iterate over each peer, removing dead/disconnected peers from the container and calling `on_data_received` for each peer

Most of the existing functions should be modified to accept a peer as a parameter, so they know which one they're working with (instead of using the global peer object we have now)

CyanBlob commented:
I was going to make a PR into this fork, but I don't seem to have access to create a new branch on your fork. So I went ahead and added it to my own fork here: https://github.com/CyanBlob/godot/commit/b8ea07ad62de0b9799bfe1f7c11dca773199f394

It seems to be working, however neither your branch nor mine work in vim (using coc.nvim):
```
[coc.nvim] error: Uncaught exception: Error: Header must provide a Content-Length property.
    at StreamMessageReader.onData (/home/andrew/.vim/plugged/coc.nvim/build/index.js:39102:27)
    at Socket.readable.on (/home/andrew/.vim/plugged/coc.nvim/build/index.js:39087:18)
    at Socket.emit (events.js:189:13)
    at addChunk (_stream_readable.js:284:12)
    at readableAddChunk (_stream_readable.js:265:11)
    at Socket.Readable.push (_stream_readable.js:220:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
```
So I'm looking into this now. But I'd appreciate insight if you have any

ofrank123 commented:
Is there any way you can show me the actual transmissions between the server and client?  In lsp-mode for emacs I can just turn lsp-print-io but I don't know anything about coc.nvim.  I'll start work on the multiple-client support, and I'll try to see why you were unable to make a fork of this fork.

CyanBlob commented:
Here's a sample:
```
[coc.nvim] error: Uncaught exception: Error: �^A^@^@Content-Length: 452^M
^M
{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":5,"message":"The argument 'event' is never used in the function '_input'. If this is intended, prefix it with an underscore: '_event'","range":{"end":{"character":19,"line":26},"start":{"character":0,"line":26}},"severity":2,"source":"gdscript"}],"uri":"file:///home/andrew/Development/deckbuilding_roguelike/deckbuilding_roguelike/Scripts/Player/Player.gd"}}
```
There's a screenshot of Wireshark output on the issue. That's the best I can do for now

With regards to the multiple clients, I got everything working pretty easily, although I wasn't sure what exactly to do when a client connects. For now, I just created an empty parameters object, which, now that I think about it, could have something to do with these malformed messages coming in. I didn't test your branch before I started changing things, so I'll double check tonight.

Feel free to copy the work from my branch or to go your own way

CyanBlob commented:
Actually, I suspect that the issue is the put_string(...) call in notify_client. It's the only place I haven't explicitly tested yet, and @Faless mentioned that put_string(...) may not be the correct method in this case

ofrank123 commented:
I'm not sure what's going with the forking, but I'm going to go ahead just implement multiple clients on my own, I think it's probably better to keep things on this fork.  I fixed the put_string(...) issue, just changed it to put_data(...).  put_string(...) prepends the size or something, Hence the weird bytes before the header.  If you could test the new version with coc.nvim to see if you have any luck that'd be great.

CyanBlob commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -189,14 +191,15 @@ void GDScriptLanguageProtocol::poll() {
 		StreamPeerTCP::Status status = peer->get_status();
 		if (status == StreamPeerTCP::STATUS_NONE || status == StreamPeerTCP::STATUS_ERROR) {
-			on_client_disconnected();
+			on_client_disconnected(c_ind);
```
Is this safe? `on_client_disconnect` modifies `clients`, which you're currently iterating over. I think that's undefined behavior

ofrank123 replied:
yeah you're right, lemme fix that

CyanBlob replied:
I tried to make it cause a problem but it refused lol. Undefined behavior is my least favorite part of C++

ofrank123 replied:
I'd rather write rust, but hey what can you do.

ofrank123 commented:
I believe this is finished, the VSCode and Atom clients will need to be changed from websockets to raw TCP.  I've made @Geequlim a reviewer because he is the primary dev on the language server code.

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -48,16 +49,16 @@ class GDScriptLanguageProtocol : public JSONRPC {
-	void on_data_received(int p_id);
-	void on_client_connected(int p_id, const String &p_protocal);
-	void on_client_disconnected(int p_id, bool p_was_clean_close);
+	void on_data_received(int c_ind, int bytes);
```
Parameters should follow godot code style using `p_` prefix for **in** and `r_` prefix for **out**.

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +188,59 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+		Ref<StreamPeerTCP> peer = clients.get(c_ind);
+		peer->disconnect_from_host();
 	}
-	server->stop();
```
Stop the tcp server here.

ofrank123 replied:
The diff is a bit weird but I stop it on 299

Geequlim commented:
Are you going to make changes to the VSCode plguin and Atom one.
If not, you'd better don't remove the websocket server.

ofrank123 commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -48,16 +49,16 @@ class GDScriptLanguageProtocol : public JSONRPC {
-	HashMap<int, Ref<WebSocketPeer> > clients;
-	WebSocketServer *server;
-	int lastest_client_id;
+	Vector<Ref<StreamPeerTCP> > clients;
```
I resolve this by queueing the clients that need to be disconnected when I poll, and then removing them after the poll step, but it'd be easy to change it to a map if you think that would work better.

Faless commented:
I tried answering by email, but appar it didn't get through, so I might be reposting this but:

"I'm traveling back from godotcon so forgive my brevity.
This code does way too many dynamic allocations, please do not use malloc (memalloc) and use a preallocated (per-peer) buffer instead (e.g. a fixed size PoolVector or array)."

ofrank123 commented:
Sounds good, I'm coming from writing mostly C and I haven't done this in a while.

Faless commented:
Still not good, use a *per peer*, *fixed size* buffer. There are examples in modules/websocket/lws_server.cpp .
Also, don't use blocking io (get_data) but get_partial_data , and don't do that while busy waiting.
Again, there are examples in the websocket server implemention.
As it is now, a malicious client could freeze/crash the engine with a malformed request.

ofrank123 commented:
Gotcha, thanks for the advice on looking at wsl_server

ofrank123 commented:
I've updated this to be safer and implemented the suggestions @Faless.  Addtionally, I have made pull requests on both the VSCode and Atom plugins to update them to work with this PR.  I have of course tested both, but I'm unfamiliar with VSCode and Atom, so it is well within the realm of possibility that I have overlooked something.  I'm rerequesting code review from @Geequlim because of the large changes, and it can now be tested on VScode and Atom.

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -31,33 +31,55 @@
-#include "modules/websocket/websocket_peer.h"
-#include "modules/websocket/websocket_server.h"
+
+#define LSP_MAX_BUFFER_SIZE 4096
```
I think the buffer size is too small.
You can try open a script large than 4KB

Shatur95 replied:
I would also suggest to use `constexpr` variable instead of macro.

Calinou replied:
@Shatur95 To my knowledge, we don't allow the use of `constexpr` in Godot's codebase. We currently only allow C++03 code to be merged; a subset of C++11 (or C++14) will soon be allowed.

Shatur95 replied:
@Calinou, bad to hear :( What is the reason of it?

Calinou replied:
@Shatur95 The reason we only plan on allowing a subset of modern C++ is to ease online code reviews (among other things). When you're a pull request reviewer, you typically don't clone people's branches in an IDE; you just rely on GitHub's code viewer which doesn't provide things like type information on hover. For instance, this is why we'll disallow the use of `auto` in Godot's codebase.

Shatur95 replied:
@Calinou, this decision is strange to me. I agree that using `auto` can really reduce readability. But this is because it does not intended to be used always. For example, in the constructions such `auto * widget = new QWidget` looks good because allows to get rid of duplicate type specifying.
`constexpr` as for me is always better than macro constant because it do the same but in safe way.

akien-mga replied:
We will use `constexpr` in the future, but this codebase is currently C++03 and introducing a stray `constexpr` when we have hundreds of `#define`s doesn't make sense. The C++14 port has to happen first.

Shatur95 replied:
@akien-mga, good point, thanks for the explanation.

ofrank123 replied:
@Geequlim any recomendations about a reasonable buffer size?  I agree that 4kb is too small.

Shatur95 commented:
Is this helps to make it work with Qt Creator and Kate?

ofrank123 commented:
@Shatur95 I'm not familiar with Qt Creator or Kate, but assuming they implement LSP clients that can connect to a TCP server (this is very common, whereas websockets are not), yes this would add support to them.  My main goal was adding support for Emacs and Vim while maintaining support for VSCode and Atom, but hopefully it will make other text editors be able to connect to the language server as well.

Shatur95 commented:
Sorry for stupid questions, I just trying to understand how it works... Both of them requires to specify path to executable server (for cases when server is not running) and do not have the ability to specify connection options. How does it work in Vim and Emacs?
QtCreator:
![изображение](https://user-images.githubusercontent.com/22453358/74154284-10542780-4c1b-11ea-8d07-96610106a60c.png)
Kate: [settings.json](https://invent.kde.org/kde/kate/blob/master/addons/lspclient/settings.json)

ofrank123 commented:
Godot's language server does not provide an executable, because the TCP server is run by the engine automatically on startup.  In the current text editors that support the GDScript LSP, they just connect to the server instead of running an executable.  Since this does not seem to be the case in kate (just going by your screenshot), you might want to try setting the executable to netcat (/usr/bin/nc) and the arguments to "localhost 6008".  I can't guarentee this will work but it may be your best bet if Kate only provides stdio support.  You'll have to make sure godot is running or nc won't be able to connect.

Shatur95 commented:
@ofrank123, thank you, it works with `nc` and your branch!
Would be nice to have CLI option to use `stdio` instead of TCP.

ofrank123 commented:
Oh great! nc was seriously a long shot.  I think that's a fine idea but a bit outside of the scope of this PR, maybe a feature request?

Shatur95 commented:
@ofrank123, done!

Faless commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +226,55 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+		} else {
+			if (peer->connection->get_available_bytes() > 0) {
+				latest_client_id = *id;
+				peer->handle_data();
```
Disconnect on errors:
```cpp
Error err = peer->handle_data();
if (err != OK && err != ERR_BUSY) {
    /// call disconnect function, free peer.
}
```

Faless commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -153,61 +226,55 @@ void GDScriptLanguageProtocol::initialized(const Variant &p_params) {
+				latest_client_id = *id;
+				peer->handle_data();
+			}
+			peer->send_data();
```
Same as above, disconnect on error.

Faless reviewed (changes requested):
This looks much much better! Great job! :1st_place_medal: !
See the few comments left for changes on server, peer disconnect handling, and my suggestion on buffer size.

Please [squash your commits](https://docs.godotengine.org/en/3.1/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) after the changes.

Again, great job!

Note: I haven't tested with a LSP client, only requests via telnet

ofrank123 commented:
Appologies for getting the docs team requested on this, a messed up the rebase.  Feel free to ignore.

ofrank123 commented:
damn it, just accidentally deleted all my progress since the first commit, then force pushed it.  Won't do that again.  I'll rewrite it tomorrow if I have time, shouldn't be too difficult hopefully.

ofrank123 commented:
Okay I think this is ready to be accepted.  Maybe a few more people could test it?

NathanLovato commented:
I have just built the branch and tested it with the Emacs LSP package, I can confirm it's working here.

rafaeldelboni commented:
I just build the branch and is perfectly working on nvim with coc.nvim!
Tested go to definition, auto complete and error linting as everyone can check on print screens bellow:
![2020-02-15_08:28:32](https://user-images.githubusercontent.com/1683898/74587111-4bfd4180-4fcd-11ea-85c5-d94c7cb0e49f.png)

![2020-02-15_08:26:12](https://user-images.githubusercontent.com/1683898/74587078-ef9a2200-4fcc-11ea-90c7-24b95f5524f5.png)

I used this configuration as reference:
https://github.com/godotengine/godot/issues/34523#issuecomment-582144661

If someone need more testing just let me know.

Faless reviewed (approved):
Looks good to me, good job! :medal_sports:

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -31,33 +31,56 @@
 	GDCLASS(GDScriptLanguageProtocol, JSONRPC)
 
+private:
+	class LSPeer : public Reference {
```
I'd like to use `struct` instead of `class` as all of the members are public and no constructor need as we can use inline initializing.

Faless replied:
It's not just a class it's reference, to avoid explicit deallocation, also, having `Vector` and `Ref` as members it is still not a `POD`, it needs an explicit initialization I think.

Faless replied:
my bad, you only meant to define it as `struct` not changing the inheritance, that make sense.

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -31,33 +31,56 @@
-	void on_client_connected(int p_id, const String &p_protocal);
-	void on_client_disconnected(int p_id, bool p_was_clean_close);
+	Error on_client_connected();
+	void on_client_disconnected(int c_ind);
```
Keep the code style p_client for the parmater name?

akien-mga replied:
Indeed, all parameters should use `p_` prefix and an explicit name (so `p_client` would be much better than `c_ind` here indeed).

Shatur95 commented:
@rafaeldelboni, do you used [this](https://github.com/clktmr/vim-gdscript3) plugin or just configured LSP in coc?

rafaeldelboni commented:
> @rafaeldelboni, do you used [this](https://github.com/clktmr/vim-gdscript3) plugin or just configured LSP in coc?

I used the old version of the plug-in, since the fork is not working anymore for me, but I'm searching for an lighter alternative just for sintax highlight and file type definition, because lsp makes all other plug-in functionalities obsolete.

Shatur95 commented:
Tested with Kate, works via netcat with this LSP config:
```json
{
    "servers": {
        "godot": {
            "command": ["nc", "localhost", "6008"],
            "url": "https://github.com/godotengine/godot",
            "highlightingModeRegex": "^GAP$"
        }
    }
}
```

rafaeldelboni commented:
@ofrank123 Can we get this merged? This is life changing, seriously 😬 
Thanks for doing this!

ofrank123 commented:
Fixed the merge conflict, as well as resolving @Geequlim's review.  Should be ready to merge.

ofrank123 commented:
I saw this was assigned the "cherrypick:3.2" flag as well.  Should I commit these changes to the 3.2 branch as well as the master branch?  I've been using 3.2 because of how unstable 4.0 is right now.

NathanLovato commented:
@ofrank123 no you should still pr to master. It's the maintainer who will cherry pick (akien I guess) to the 3.2 branch

akien-mga commented:
It looks like you have an extra merge commit in this PR. Could you rebase on upstream master to squash it together with the main commit?

ofrank123 commented:
Squashed!

akien-mga commented:
Just FYI, the email address used to author the commit (see https://github.com/godotengine/godot/commit/24b27043fec7ef9a4baf2c5ecab67fc20563b586.patch) is not associated with your @ofrank123 GitHub account, which is why the commit appears authored and committed by two different persons. It's not a problem for Git nor for merging this PR, but if you want to solve it, you can either edit the commit author or add this email as secondary email in your GitHub settings.

ofrank123 commented:
Huh, weird.  Should be fixed.

akien-mga commented:
Thanks!

akien-mga commented:
#36485 should be cherry-picked together with this PR.

Shatur95 commented:
This PR was not cherry-picked in 3.2.1 release?

akien-mga commented:
No, it's scheduled for 3.2.2.

rafaeldelboni commented:
May I ask when 3.2.2 is planned to be released?

I hope all LSP merged prs will be in this one 😬

akien-mga commented:
> May I ask when 3.2.2 is planned to be released?

When it's ready.

> I hope all LSP merged prs will be in this one grimacing

That's the plan yes.

CyanBlob commented:
This doesn't seem to be working for me on master using vim (coc.nvim plugin). Is anyone seeing something similar?
`[Error  - 5:36:10 PM] Connection to server is erroring. Shutting down server.`

rafaeldelboni commented:
I be using this since the merge with nvim I and It works quite well to me

CyanBlob commented:
Weird. It seems to only affect the `release_debug` build. I rebuilt `debug` and it works

rafaeldelboni commented:
Ah maybe is something in the `master` branch, I'm using the current `3.2` with #35864, #36472, #36521 and #36683 cherypicked.

Shatur95 commented:
@akien-mga, could you cherry-pick it to the 3.2 branch to simplify manual intervention?

VictorRogers commented:
@rafaeldelboni I'm wanting to get this working, but when I start cherry-picking commits to the 3.2 branch I hit merge conflicts pretty fast. Did you have to manually resolve them?

rafaeldelboni commented:
> @rafaeldelboni I'm wanting to get this working, but when I start cherry-picking commits to the 3.2 branch I hit merge conflicts pretty fast. Did you have to manually resolve them?
> 
> For example, when merging #35864 I get merge conflicts in the gdscript_language_protocol.cpp file.

Try using a strategy option to do this cherry-pick:
`git cherry-pick 24b27043fec7ef9a4baf2c5ecab67fc20563b586 --strategy-option theirs`

VictorRogers commented:
@rafaeldelboni Will do! Thanks

VictorRogers commented:
@rafaeldelboni Looks like I'm making some progress. When I fire up vim with coc.nvim, I get the following error:
[coc.nvim] Server languageserver.godot failed to start: Connection error for languageserver.godot: connect ECONNREFUSED 127.0.0.1:6008

NathanLovato commented:
For those who want to compile the 3.2 branch with the LSP support, you can run the following code in your shell from Godot's repository (bash, sh):

```sh
git checkout 3.2
for i in 24b27043fec7ef9a4baf2c5ecab67fc20563b586 ed482f6167276bebd4132b82a6683392ac41f97a 06bce137e38ecf0e3c024ddff8356c8cfd9c4d5b 2f08f4ef4e6f7112c1c06f9f6b6f7e782e66adaf; do
    git cherry-pick $i --strategy-option theirs
done
```

Then rebuild the engine.

akien-mga commented:
Cherry-picked for 3.2.2.

aaronfranke commented:
Why was this cherrypicked? It may cause issues with the VS Code plugin: https://github.com/godotengine/godot-vscode-plugin/issues/154

NathanLovato commented:
The language server implementation wouldn't work on pretty much any other editor than vscode and atom as the vast majority of language servers out there use TCP - while for the vscode and atom plugin, it's a matter of moving the transport method from websocket to tcp.

Also, the language server shipped in 3.2 wasn't following the specification in some areas - it's still the case as far as I know.

pavanscoding commented:
> Disregard below. Restarted godot and it's working now. Thanks! I'll post back with any issues I find.
> 
> @rafaeldelboni Looks like I'm making some progress. When I fire up vim with coc.nvim, I get the following error: [coc.nvim] Server languageserver.godot failed to start: Connection error for languageserver.godot: connect ECONNREFUSED 127.0.0.1:6008

I am also getting the same error while using coc.nvim along with godot. I was wondering if you ever found a solution to this error?

Calinou commented:
> I am also getting the same error while using coc.nvim along with godot. I was wondering if you ever found a solution to this error?

Which Godot version are you using? Most extensions aren't compatible with 4.0.alpha yet.
